ts/client/src/accounts/mangoAccount.ts

```diff
@@ -1628,6 +1676,243 @@ export class PerpPosition {
+            I80F48.fromI64(a.sizeLots)
+              .mul(I80F48.fromI64(perpMarket.baseLotSize))
+              .mul(perpMarket.priceLotsToNative(a.priceLots)),
+          ),
+        ZERO_I80F48(),
+      ),
+    };
+  }
+
+  public getLiquidationPriceInternal(
+    group: Group,
+    perpMarket: PerpMarket,
+    mangoAccount: MangoAccount,
+    sortedOo: PerpOrder[],
+    tillPriceExclusive: I80F48,
+  ): I80F48 {
+    const settleTokenBank = group.getFirstBankByTokenIndex(
+      perpMarket.settleTokenIndex,
+    );
+    const settleTokenPrice = settleTokenBank.price;
+
+    // For constants B, T, C and final liqudation price equation, refer to
+    // https://www.notion.so/mango-markets/Mango-v4-Perp-Liq-Price-b7dde5e70c724d07934a21dc0aa79518?pvs=4
+
+    // Set base and quote
+    // Take into account bids or asks if they'd execute
+    // before we hit tillPriceExclusive i.e. previously computed liquidation price
+    let wbpn; // B
+    let qpn = this.quotePositionNative.sub(
+      this.getUnsettledFunding(perpMarket),
+    );
+    {
+      const baseNative = this.getBasePositionNative(perpMarket);
+      if (baseNative.isZero()) {
+        return MINUS_ONE_I80F48();
+      }
+
+      if (baseNative.isNeg()) {
+        const ret = PerpPosition.getAggregatePerpOrderSizeAndQuoteNative(
+          perpMarket,
+          sortedOo,
+          PerpOrderSide.ask,
+          tillPriceExclusive,
+        );
+        wbpn = perpMarket.maintBaseLiabWeight.mul(
+          baseNative.sub(
+            I80F48.fromI64(ret.sizeLots).mul(
+              I80F48.fromI64(perpMarket.baseLotSize),
+            ),
+          ),
+        );
+        qpn = qpn.add(ret.quoteNative);
+      } else {
+        const ret = PerpPosition.getAggregatePerpOrderSizeAndQuoteNative(
+          perpMarket,
+          sortedOo,
+          PerpOrderSide.ask,
+          tillPriceExclusive,
+        );
+        wbpn = perpMarket.maintBaseAssetWeight.mul(
+          baseNative.add(
+            I80F48.fromI64(ret.sizeLots).mul(
+              I80F48.fromI64(perpMarket.baseLotSize),
+            ),
+          ),
+        );
+        qpn = qpn.sub(ret.quoteNative);
+      }
+    }
+
+    if (wbpn.isZero()) {
+      return MINUS_ONE_I80F48();
+    }
+
+    // Health contribution of perp position in question would be considered separately
+    const mAClone: MangoAccount = cloneDeep(mangoAccount);
+    const ppToBeErased: PerpPosition = mAClone.getPerpPosition(
+      perpMarket.perpMarketIndex,
+    )!;
+    ppToBeErased.marketIndex =
+      PerpPosition.PerpMarketIndexUnset as PerpMarketIndex;
+    const healthCacheWoPerpPosition = HealthCache.fromMangoAccount(
+      group,
+      mAClone,
+    );
+
+    // Unused, TODO what to do with this?
+    // const basePrice = qpn.neg().div(wbpn);
+
+    const healthContributionWithoutSettleToken =
+      healthCacheWoPerpPosition.healthExcludingToken(
+        HealthType.maint,
+        perpMarket.settleTokenIndex,
+      ); // C
```

Comment:
For this, you don't need the WoPerpPosition, since you exclude the whole token index anyway. If you drop the right perp market in the filter() in line 1858, you might not need to manually wipe the perp pos at all.